pygfx/objects/_ruler.py

```diff
@@ -0,0 +1,489 @@
+        length = end_value - start_value
+        vec = end_pos - start_pos
+        if length:
+            vec /= length
+
+        # Get array to store positions
+        n_slots = self.points.geometry.positions.nitems
+        n_positions = len(ticks) + 2
+        if n_positions <= n_slots <= max(min_n_slots, 2 * n_positions):
+            # Re-use existing buffers
+            positions = self.points.geometry.positions.data
+            sizes = self.points.geometry.sizes.data
+            self.points.geometry.positions.update_range()
+            self.points.geometry.sizes.update_range()
+        else:
+            # Allocate new buffers
+            new_n_slots = max(min_n_slots, int(n_positions * 1.2))
+            positions = np.zeros((new_n_slots, 3), np.float32)
+            sizes = np.zeros((new_n_slots,), np.float32)
+            self.points.geometry.positions = Buffer(positions)
+            self.points.geometry.sizes = Buffer(sizes)
+
+        def define_text(pos, text):
+            if index < len(self._text_object_pool):
+                ob = self._text_object_pool[index]
+            else:
+                ob = Text(
+                    TextGeometry("", screen_space=True),
+                    TextMaterial(),
+                )
+                self._text_object_pool.append(ob)
+            ob.geometry.anchor = self._text_anchor
+            ob.geometry.anchor_offset = self._text_anchor_offset
+            ob.geometry.set_text(text)
+            ob.local.position = pos
+
+        # Apply start point
+        index = 0
+        positions[0] = start_pos
+        if self._ticks_at_end_points:
+            sizes[0] = tick_size
+            define_text(start_pos, f"{self._start_value:0.4g}")
+        else:
+            sizes[0] = 0
+            define_text(start_pos, f"")
+
+        # Collect ticks
+        index += 1
+        for value, text in ticks.items():
+            pos = start_pos + vec * (value - start_value)
+            positions[index] = pos
+            sizes[index] = tick_size
+            define_text(pos, text)
+            index += 1
+
+        # Handle end point
+        positions[index:] = end_pos
+        sizes[index:] = 0
+        self._text_object_pool[index + 1 :] = []
+
+        if self._ticks_at_end_points:
+            sizes[index] = tick_size
+            define_text(end_pos, f"{end_value:0.4g}")
+        else:
+            define_text(end_pos, "")
+
+        # Hide the ticks close to the ends?
+        if self._ticks_at_end_points and ticks:
+            tick_values = list(ticks.keys())
+            if abs(tick_values[0] - start_value) < 0.5 * tick_auto_step:
+                self._text_object_pool[1].geometry.set_text("")
+            if abs(tick_values[-1] - end_value) < 0.5 * tick_auto_step:
+                self._text_object_pool[index - 1].geometry.set_text("")
+
+        # Apply
+        self.clear()
+        self.add(self._line, self._points)
+        self.add(*self._text_object_pool)
```

Comment:
does this mean the same objects are added over and over again?

Reply:
Good point, we can limit adding/removing stuff to when the pool is resized. Will do.